MK3: TM: Prevent lockout on invalid model values

wavexx opened the pull request:
If the model is changed to invalid values (by manually running ``M310 C0.01`` for example) the printer will enter a continuous state of thermal error where 1) the LCD is unusable and 2) the serial is disabled. This cannot be undone except for a factory reset.

Allow again to use the menus after a thermal error. Since the thermal model is managing the heaters, the user can't worsen the situation. This can be used to run gcode from the SD.

Allow M310 to bypass the serial lock. M310 doesn't control heaters, and can't be used to bypass the stopped state, so it's safe to run in an emergency situation to reset the model.

~This is still incomplete though: if the thermal anomaly is constantly triggering (as happens with wrong values), even though the menus are unlocked, the error will continuously force redisplay, preventing you from doing anything.~

- [X] Create cherry-pick PR to [MK3_3.12 branch](https://github.com/prusa3d/Prusa-Firmware/pull/3845)

wavexx committed:
TM: Do not lockout the menus on thermal errors

This prevents the ability to run gcode from the SD card.

In a thermal error with faulty values, and without serial access, this
is the only way to process an M310 instruction and recover.

wavexx committed:
Keep a pointer past the line number in cmdqueue

This allows to skip line numbers transparently and perform faster
checks on the actual command to process.

wavexx committed:
Allow M310 to bypass the Stopped state for recovery

Since M310 cannot change the heaters, allowing M310 S0 (and changing
parameters) allows to recover a usb-controlled printer which has been
locked-out due to bad model settings.

3d-gussner reviewed (changes requested):
Please review and add https://github.com/wavexx/Prusa-Firmware/pull/10

Tested this on MK404 which does directly into TM Error.
- Sending `M310 S0` :heavy_check_mark: 
  - But after reset, which is the only option printer is still locked
    - With the PR above I am able to disable the temp model and save it, so next time it reboots it stays disabled.

DRracer commented on Firmware/cmdqueue.cpp:
```diff
@@ -422,25 +424,24 @@ void get_command()
 		}
         // if we don't receive 'N' but still see '*'
-        if ((cmdbuffer[bufindw + CMDHDRSIZE] != 'N') && (cmdbuffer[bufindw + CMDHDRSIZE] != '$') && (strchr(cmdbuffer+bufindw+CMDHDRSIZE, '*') != NULL))
+        if ((*cmd_head != 'N') && (cmdbuffer[bufindw + CMDHDRSIZE] != '$') && (strchr(cmd_start, '*') != NULL))
```
Why wasn't `cmdbuffer[bufindw + CMDHDRSIZE]` also changed to `*cmd_head`? I don't see a difference yet...

wavexx replied:
Probably because I removed it entirely on #3823. I will fix this

DRracer commented on Firmware/cmdqueue.cpp:
```diff
@@ -422,25 +424,24 @@ void get_command()
-                usb_timer.start();
-            }
-        }
+        if ((*cmd_start == 'G') && !IS_SD_PRINTING)
```
suggest parentheses around `IS_SD_PRINTING` as it is a macro and unintentional change of it may cause hard-to-debug issues in the future

DRracer reviewed (approved):
Looks correct to me, just 2 minor stylistic questions. Otherwise I don't see any further issues with these changes

wavexx committed:
Replace two expressions with cmd_head

wavexx committed:
Parenthesize IS_SD_PRINTING defensively

wavexx commented:
Regarding https://github.com/wavexx/Prusa-Firmware/pull/10 I would handle it differently. We need to allow the user to unlock the printer from the LCD.

Besides adding code in a hot path, M500 can save garbage state, and allowing the command there will make it available also while breaking the paused protocol, something I'm really not keen on.

wavexx committed:
Rename print_stop() to lcd_print_stop_finish() for clarity

wavexx committed:
TM: Clear the Stopped state when stopping the current print

We allow resuming from the LCD via start print and resume print,
it makes sense to clear the error on stop too.

For this reason distinguish whether the action is performed
automatically or manually (ie: interactively).

The error is only cleared when the command is run interactively.

wavexx committed:
TM: correctly mark repeated anomalies

wavexx committed:
Do not show/call LCD status updates when unchanged

This mostly prevents useless serial noise

wavexx committed:
Use menu_set_block() in ThermalStop()

Set a menu block for fatal thermal errors instead of abusing
lcd_return_to_status() to kick the user out of the menus.

This now allows a thermal model error to be recoverable through menu
access.

wavexx committed:
TM: Allow to resume without a running print

If there is no running print, and the printer is Stopped, add a new
"Acknowledge error" menu entry to unlock the printer.

This simply calls lcd_print_stop(), which is identical in behavior to a
thermal error with a running print.

wavexx committed:
TM: Disallow preheat/filament actions when stopped

Require acknowledgement first

wavexx commented:
We the latest commits we do the following:

- Repeated thermal model warnings/errors now do not prevent menu access
- Only partial menu access is available: preheat/filament/calibrate actions are disabled
- The user can clear a thermal error by using "Acknowledge error". This works for a one-time error, and will resume USB serial processing as expected in the same way aborting/resuming a thermal error pause does.
- "Print from SD" is now available during a thermal error, and a new print will implicitly *clear* the error state. This is allowed as an escape hatch to run g-code containing "M310 S0" to disable the model. If the model is still running, the SD print will still be interrupted pretty fast, so it should pose no threat.

wavexx committed:
Also prevent longpress when Stopped

3d-gussner commented on Firmware/messages.cpp:
```diff
@@ -169,6 +169,7 @@ extern const char MSG_PAUSED_THERMAL_ERROR[] PROGMEM_I1 = ISTR("PAUSED THERMAL E
 #ifdef TEMP_MODEL
 extern const char MSG_THERMAL_ANOMALY[] PROGMEM_I1 = ISTR("THERMAL ANOMALY");////MSG_THERMAL_ANOMALY c=20
 extern const char MSG_TM_NOT_CAL[] PROGMEM_I1 = ISTR("Temp model not calibrated yet.");////MSG_TM_NOT_CAL c=20 r=4
+extern const char MSG_ACK_ERROR[] PROGMEM_I1 = ISTR("Acknowledge error");////MSG_ACK_ERROR c=17
```
A menu can be `c=18`
The message may not be that clear to the users.
```
01234567890123456789
>Clear TM error    ➔
>Confirm TM error  ➔
>Unlock TM error   ➔

wavexx committed:
Improve MSG_ACK_ERROR message

wavexx commented:
This fixes #3630

3d-gussner commented on Firmware/messages.cpp:
```diff
@@ -169,6 +169,7 @@ extern const char MSG_PAUSED_THERMAL_ERROR[] PROGMEM_I1 = ISTR("PAUSED THERMAL E
 #ifdef TEMP_MODEL
 extern const char MSG_THERMAL_ANOMALY[] PROGMEM_I1 = ISTR("THERMAL ANOMALY");////MSG_THERMAL_ANOMALY c=20
 extern const char MSG_TM_NOT_CAL[] PROGMEM_I1 = ISTR("Temp model not calibrated yet.");////MSG_TM_NOT_CAL c=20 r=4
+extern const char MSG_TM_ACK_ERROR[] PROGMEM_I1 = ISTR("Clear TM error");////MSG_TM_ACK_ERROR c=18
```
Adding new message needs an update of the pot and po files.
Also the new message should be translated.
Do want to do that in a separate PR?
Splitting "unblocks" this PR as we don't need to wait for the translations, which may take a while.

wavexx replied:
Why would it block it?

3d-gussner replied:
Because we have to wait for the translations

3d-gussner reviewed (approved):
Works well on MK404 and real printer.
On MK404 after factory reset I am able to recover from TM error and "unlock" the printer. This solves the worst case scenario of locking out users.
